### PR TITLE
Developer Guide: SDK Utilities

### PR DESCRIPTION
**Pull request by skmcgrail**
Splits and updates the content of `content/en/sdk-utilities.md` into a folder hierarchy under `content/en/sdk-utilities/`.

### CODE REVIEW

**Review by smckenzie23 — APPROVED** [review]

File: content/en/docs/sdk-utilities/s3.md [hunk 1]
```diff
@@ -0,0 +1,430 @@
...
+		}
+		close(walker)
+	}()
+
+	cfg, err := config.LoadDefaultConfig()
+	if err != nil {
+		log.Fatalln("error:", err)
+	}
+	
+	// For each file found walking, upload it to S3
```
Comment (Doug-AWS) [inline content/en/docs/sdk-utilities/s3.md:204]: Amazon S3

File: content/en/docs/sdk-utilities/s3.md [hunk 2]
```diff
@@ -0,0 +1,430 @@
...
+```
+
+## GetBucketRegion
+
+The [GetBucketRegion]({{% apiref "feature/s3/manager#GetBucketRegion" %}}) is a utility function for determining the
+AWS Region location of an {{% alias service=S3 %}} Bucket. `GetBucketRegion` takes an {{% alias service=S3 %}}
+client and uses it to determine the location of the requested Bucket within the AWS Partition associated with the
+client's configured Region.
+
+For example to find the region the Bucket `my-bucket`:
```
Comment (Doug-AWS) [inline content/en/docs/sdk-utilities/s3.md:405]: Region

File: content/en/docs/sdk-utilities/s3.md [hunk 3]
```diff
@@ -0,0 +1,430 @@
...
+	log.Println("error:", err)
+	return
+}
+
+bucket := "my-bucket"
+region, err := manager.GetBucketRegion(ctx, s3.NewFromConfig(cfg), bucket)
+if err != nil {
+	var bnf manager.BucketNotFound
+	if errors.As(err, &bnf) {
+		log.Printf("unable to find bucket %s's region\n", bucket)
```
Comment (Doug-AWS) [inline content/en/docs/sdk-utilities/s3.md:419]: Region

File: content/en/docs/sdk-utilities/s3.md [hunk 4]
```diff
@@ -0,0 +1,430 @@
...
+	} else {
+		log.Println("error:", err)
+	}
+	return
+}
+fmt.Printf("Bucket %s is in %s region\n", bucket, region)
+```
+
+If `GetBucketRegion` is not able to resolve the location of a Bucket, the function will return a
```
Comment (Doug-AWS) [inline content/en/docs/sdk-utilities/s3.md:428]: will return -> returns

**Review by Doug-AWS — CHANGES_REQUESTED** [review]

File: content/en/docs/sdk-utilities/s3.md [hunk 5]
```diff
@@ -0,0 +1,430 @@
...
+The `Concurrency` value limits the concurrent number of part uploads that can occur for a given `Upload` call. 
+This is not a global client concurrency limit. Tweak the `PartSize` and `Concurrency` configuration values to find the
+optimal configuration. For example, systems with high-bandwidth connections can send bigger parts and more uploads in parallel.
+
+For example, your application configures `Uploader`  with a `Concurrency` of setting of `5`. Your application then 
+calls `Upload` from two different goroutines, the result will be `10` concurrent part uploads
+(2 gorutines * 5 `Concurrency`).
+
+{{% pageinfo color="warning" %}}
+Your application is expected to limit the concurrent calls to `Upload` to prevent application resource exhaustion.
```
Comment (Jerry-AWS) [inline content/en/docs/sdk-utilities/s3.md:62]: A little more detail on what a sensible limit would be, or how to determine one, would be welcome.

File: content/en/docs/sdk-utilities/s3.md [hunk 6]
```diff
@@ -0,0 +1,430 @@
...
+         // Process error generically
+         fmt.Println("Error:", err.Error())
+     }
+ }
+```
+
+### Overriding Uploader Options Per Upload
+
+You can override the `Uploader` options when calling `Upload` by providing one or more functional arguments to the
+method. These overrides are concurrency safe modifications and do not affect ongoing uploads, or subsequent `Upload`
```
Comment (Jerry-AWS) [inline content/en/docs/sdk-utilities/s3.md:137]: **concurrency safe** -> **concurrency-safe**

File: content/en/docs/sdk-utilities/s3.md [hunk 7]
```diff
@@ -0,0 +1,430 @@
...
+	Key:    aws.String("my-key"),
+})
+```
+
+The `downloadFile` parameter is an `io.WriterAt` type. The `WriterAt` interface enables the `Downloader` to write
+multiple parts of the file in parallel.
+
+#### Configuration Options
+
+When you instantiate a `Downloader` instance, you can specify several configuration options to customize how objects are
```
Comment (Jerry-AWS) [inline content/en/docs/sdk-utilities/s3.md:280]: Think "several" might be overstating things here; looks like there are only two. Could just omit "several."

**Review by Jerry-AWS — APPROVED** [review]
only minor changes, none of which are essential

File: content/en/docs/sdk-utilities/s3.md [hunk 8]
```diff
@@ -0,0 +1,430 @@
...
+})
+```
+
+### Configuration Options
+
+When you instantiate an `Uploader` instance using [NewUploader]({{% apiref "feature/s3/manager#NewUploader" %}}), you
+can specify several configuration options to customize how objects are uploaded. Options are overridden by providing one
+or more functional arguments to `NewUploader`. These options include:
+
+* `PartSize` &ndash; Specifies the buffer size, in bytes, of each part to upload. The minimum size per part is 5 MB.
```
[the remark this answers is not in the document]
Comment (skmcgrail) [reply content/en/docs/sdk-utilities/s3.md:49]: Good call out, for other occurrences in the code I have a comment I call out this ([MiB](https://en.wikipedia.org/wiki/Mebibyte)) which is the base-2 value. so 5242880.

File: content/en/docs/sdk-utilities/s3.md [hunk 9]
```diff
@@ -0,0 +1,430 @@
...
+   Bucket: aws.String("my-bucket"),
+   Key:    aws.String("my-object-okey"),
+   Body:   uploadFile,
+})
+ output, err := u.upload(input)
+ if err != nil {
+ 	var mu manager.MultiUploadFailure
+     if errors.As(err, &mu) {
+         // Process error and its associated uploadID
+         fmt.Println("Error:", mu.Error())
```
Comment (jasdel) [inline content/en/docs/sdk-utilities/s3.md:125]: nit, `.Error()` method call not needed. and spacing is a bit off.

**Review by jasdel — APPROVED** [review]

File: content/en/docs/sdk-utilities/s3.md [hunk 10]
```diff
@@ -0,0 +1,430 @@
...
+can be split into smaller parts and downloaded in parallel. You can customize the number of parallel downloads and the
+size of the downloaded parts.
+
+#### Example: Download a File
+
+The following example uses the {{% alias service="S3" %}} `Downloader` to download a file. Using `Downloader` is similar
+to the [s3.GetObject]({% apiref "service/s3#Client.GetObject" %}}) operation.
+
+```go
+import "github.com/aws/aws-sdk-go-v2/config"
```
Comment (skotambkar) [inline content/en/docs/sdk-utilities/s3.md:254]: nit: import `aws`

File: content/en/docs/sdk-utilities/s3.md [hunk 11]
```diff
@@ -0,0 +1,430 @@
...
+   parallel.
+
+The `Concurrency` value limits the concurrent number of part download that can occur for a given `Download` call.
+This is not a global client concurrency limit. Tweak the `PartSize` and `Concurrency` configuration values to find the
+optimal configuration. For example, systems with high-bandwidth connections can recieve bigger parts and more downloads
+in parallel.
+
+For example, your application configures `Downloader`  with a `Concurrency` of `5`. Your application then
+calls `Download` from two different goroutines, the result will be `10` concurrent part downloads
+(2 gorutines * 5 `Concurrency`).
```
Comment (skotambkar) [inline content/en/docs/sdk-utilities/s3.md:295]: nit: gorutines => goroutines

**Pull request merged**